Source/MLXFast/MLXFast.swift

```diff
@@ -0,0 +1,65 @@
+// Copyright © 2024 Apple Inc.
+
+import Cmlx
+import MLX
+
+/// Optimized implementation of `NN.RoPE`.
+///
+/// Used like this:
+///
+/// ```swift
+/// let x: MLXArray
+/// let dimensions: Int
+/// let traditional: Bool
+/// let base: Float
+/// let scale: Float
+/// let offset: Int
+///
+/// let shape = x.shape
+/// var x = x.reshaped(-1, x.dim(-2), x.dim(-1))
+/// x = MLXFast.RoPE(x, dimensions: dimensions, traditional: traditional, base: base, scale: scale, offset: offset)
+/// return x.reshaped(shape)
+/// ```
+///
+/// > Note: `MLXNN.RoPE` uses this implementation internally.
+public func RoPE(
+    _ array: MLXArray, dimensions: Int, traditional: Bool, base: Float, scale: Float, offset: Int,
+    stream: StreamOrDevice = .default
+) -> MLXArray {
+    MLXArray(
+        mlx_fast_rope(
+            array.ctx, Int32(dimensions), traditional, base, scale, Int32(offset), stream.ctx))
+}
+
+/// A fast implementation of multi-head attention: `O = softmax(Q @ K.T, dim=-1) @ V`
+///
+/// Supports [Multi-Head Attention](https://arxiv.org/abs/1706.03762), [Grouped Query Attention](https://arxiv.org/abs/2305.13245), and [Multi-Query Attention](https://arxiv.org/abs/1911.02150).
+///
+/// This function will dispatch to an optimized Metal kernel when the query sequence length is 1. It handles other cases with regular MLX operations.
+///
+/// > Note: The softmax operation is performed in float32 precision regardless of input precision (float16 or float32).
+///
+/// > Note: For Grouped Query Attention and Multi-Query Attention, the input arrays for `key` and `value` should not be pre-tiled to match the `query` array.
+///
+/// Specifically this implements:
+///
+/// ```swift
+/// var scores = (queries * self.scale).matmul(keys.transposed(0, 1, 3, 2))
+/// if let mask {
+///     scores = scores + mask
+/// }
+///
+/// scores = softMax(scores.asType(.float32), axis: -1).asType(scores.dtype)
+///
+/// return matmul(scores, values).transposed(0, 2, 1, 3)
+/// ```
+///
+/// > Note: users must call this when appropriate.  `mlx-swift-examples` has been updated to call this.
```

Comment:
I don't think we need the comment re `mlx-swift-examples` in the source here.. doesn't add much IMO.